components/style/stylist.rs

```diff
@@ -98,15 +98,24 @@ impl DocumentCascadeData {
     {
         debug_assert!(!flusher.nothing_to_do());
 
-        let origins_to_rebuild = flusher.origins_to_fully_rebuild();
+        for (cascade_data, origin) in self.per_origin.iter_mut_origins() {
+            let validity = flusher.origin_validity(origin);
 
-        for origin in origins_to_rebuild.iter() {
-            extra_data.borrow_mut_for_origin(&origin).clear();
-            self.per_origin.borrow_mut_for_origin(&origin).clear();
-        }
+            if validity == OriginValidity::Valid {
+                continue;
+            }
+
+            if origin == Origin::UserAgent {
+                self.precomputed_pseudo_element_decls.clear();
+            }
 
-        if origins_to_rebuild.contains(Origin::UserAgent.into()) {
-            self.precomputed_pseudo_element_decls.clear();
+            extra_data.borrow_mut_for_origin(&origin).clear();
+            if validity == OriginValidity::CascadeInvalid {
+                cascade_data.clear_cascade_data()
```

Comment:
Yeah, I don't have any better naming idea though... I can rename the current `CascadeData` to `StyleData`, and divide it in `CascadeData` and `InvalidationData` I guess... But that seemed a bit overkill, I'll file a followup.